pennylane/capture/base_interpreter.py

```diff
@@ -0,0 +1,415 @@
+        so the output will not effect later equations in the circuit.
+
+        See also: :meth:`~.interpret_operation_eqn`.
+
+        """
+        return op._unflatten(*op._flatten())  # pylint: disable=protected-access
+
+    def interpret_operation_eqn(self, eqn: "jax.core.JaxprEqn"):
+        """Interpret an equation corresponding to an operator.
+
+        Args:
+            eqn (jax.core.JaxprEqn): a jax equation for an operator.
+
+        See also: :meth:`~.interpret_operation`.
+
+        """
+
+        invals = (self.read(invar) for invar in eqn.invars)
+        with qml.QueuingManager.stop_recording():
+            op = eqn.primitive.impl(*invals, **eqn.params)
+        if isinstance(eqn.outvars[0], jax.core.DropVar):
+            return self.interpret_operation(op)
+
+        self._op_math_cache[eqn.outvars[0]] = op
+        return op
+
+    def interpret_measurement_eqn(self, primitive, *invals, **params):
+        """Interpret an equation corresponding to a measurement process.
+
+        Args:
+            primitive (jax.core.Primitive): a jax primitive corresponding to a measurement.
+            *invals (Any): the positional input variables for the equation
+
+        Keyword Args:
+            **params: The equations parameters dictionary
+
+        """
+        invals = (
+            self.interpret_operation(op) for op in invals if isinstance(op, qml.operation.Operator)
+        )
+        return primitive.bind(*invals, **params)
+
+    def eval(self, jaxpr: "jax.core.Jaxpr", consts: list, *args) -> list:
+        """Evaluate a jaxpr.
+
+        Args:
+            jaxpr (jax.core.Jaxpr): the jaxpr to evaluate
+            consts (list[TensorLike]): the constant variables for the jaxpr
+            *args (tuple[TensorLike]): The arguments for the jaxpr.
+
+        Returns:
+            list[TensorLike]: the results of the execution.
+
+        """
+        self._env = {}
+        self._op_math_cache = {}
+        self.setup()
+
+        for arg, invar in zip(args, jaxpr.invars):
+            self._env[invar] = arg
+        for const, constvar in zip(consts, jaxpr.constvars):
+            self._env[constvar] = const
+
+        for eqn in jaxpr.eqns:
+
+            custom_handler = self._primitive_registrations.get(eqn.primitive, None)
+            if custom_handler:
+                invals = [self.read(invar) for invar in eqn.invars]
+                outvals = custom_handler(self, *invals, **eqn.params)
+            elif isinstance(eqn.outvars[0].aval, AbstractOperator):
+                outvals = self.interpret_operation_eqn(eqn)
+            elif isinstance(eqn.outvars[0].aval, AbstractMeasurement):
+                invals = [self.read(invar) for invar in eqn.invars]
+                outvals = self.interpret_measurement_eqn(eqn.primitive, *invals, **eqn.params)
+            else:
+                invals = [self.read(invar) for invar in eqn.invars]
```

Comment:
If `interpret_measurement_eqn` is not adapted to `interpret_operation_eqn` but vice versa, one could pull `outvals = [...]` out of the loop :)

Reply:
One could also put the loop body into an `interpret_eqn` function that an interpreter dev can overwrite. This could come in handy whenever handling ops, measurements etc. is  all the same (because it's just being printed/drawn or what not)

Reply:
I think we have to find a balancing point between the amount of customizability and the complexity of the class.  We could add more options to customize more components, but each time we do so, the class gets much more complicated.

If this framework still isn't working for a given situation, we can always just write our own beposke version with greater customization.  This is to be an aid to help organize the majority of the situations.

Reply:
Makes sense! :+1:

Reply:
I'm unresolving this conversation because I agree with David. I'm prototyping an implementation where I need to use `self.read` conditionally, and it doesn't work for measurements unless I mostly copy the entire `eval` function and remove the `self.read` step

Reply:
Splitting it into a `interpret_measurement_eqn(eqn)` and `interpret_measurement(measurement)` methods.